crypto/src/main/java/org/cryptimeleon/incentive/crypto/proof/spend/zkp/SpendDeductBooleanZkp.java

```diff
@@ -0,0 +1,154 @@
+    }
+
+    /**
+     * Function used to recursively generate the protocol tree.
+     */
+    private ProtocolTree generateProtocolTree(SpendDeductTree spendDeductTree, CommonInput commonInput, SendFirstValue sendFirstValue) {
+        if (spendDeductTree instanceof SpendDeductOrNode) {
+            SpendDeductOrNode spendDeductOrNode = (SpendDeductOrNode) spendDeductTree;
+            return or(
+                    generateProtocolTree(spendDeductOrNode.left, commonInput, sendFirstValue),
+                    generateProtocolTree(spendDeductOrNode.right, commonInput, sendFirstValue)
+            );
+        } else if (spendDeductTree instanceof SpendDeductAndNode) {
+            SpendDeductAndNode spendDeductAndNode = (SpendDeductAndNode) spendDeductTree;
+            return and(
+                    generateProtocolTree(spendDeductAndNode.left, commonInput, sendFirstValue),
+                    generateProtocolTree(spendDeductAndNode.right, commonInput, sendFirstValue)
+            );
+        } else if (spendDeductTree instanceof SpendDeductLeafNode) {
+            SpendDeductLeafNode spendDeductLeafNode = (SpendDeductLeafNode) spendDeductTree;
+            return leaf(
+                    spendDeductLeafNode.getLeafName(),
+                    getProtocolForLeaf(spendDeductLeafNode, pp, promotionParameters, providerPublicKey),
+                    commonInput
+            );
+        } else {
+            throw new RuntimeException("Unexpected instance of SpendDeductTree found!");
+        }
+    }
+
+    /**
+     * Wrapper function around the recursive function with the same name for adding all known witnesses to the builder.
+     */
+    private void putWitnesses(CommonInput commonInput, SecretInput secretInput, ProverSpecBuilder builder) {
+        putWitnesses(this.spendDeductTree, commonInput, secretInput, builder);
+    }
+
+    /**
+     * Recursively adds witness to all nodes that are valid the builder.
+     */
+    private void putWitnesses(SpendDeductTree spendDeductTree, CommonInput commonInput, SecretInput secretInput, ProverSpecBuilder builder) {
+        if (spendDeductTree instanceof SpendDeductOrNode) {
+            SpendDeductOrNode spendDeductOrNode = (SpendDeductOrNode) spendDeductTree;
+            putWitnesses(spendDeductOrNode.left, commonInput, secretInput, builder);
+            putWitnesses(spendDeductOrNode.right, commonInput, secretInput, builder);
+        } else if (spendDeductTree instanceof SpendDeductAndNode) {
+            SpendDeductAndNode spendDeductAndNode = (SpendDeductAndNode) spendDeductTree;
+            putWitnesses(spendDeductAndNode.left, commonInput, secretInput, builder);
+            putWitnesses(spendDeductAndNode.right, commonInput, secretInput, builder);
+        } else if (spendDeductTree instanceof SpendDeductLeafNode) {
+            SpendDeductLeafNode spendDeductLeafNode = (SpendDeductLeafNode) spendDeductTree;
+            if (spendDeductLeafNode.hasWitness()) {
```

Comment:
Replace this with asking the node whether it is fulfilled for a given point vector.

Reply:
Done